Lib/test/test_annotationlib.py

```diff
@@ -1675,6 +1675,24 @@ def test_name_lookup_without_eval(self):
 
         self.assertEqual(exc.exception.name, "doesntexist")
 
+    def test_evaluate_undefined_generic(self):
+        # Test the codepath where have to eval() with undefined variables.
+        generic = ForwardRef("glob[int, undef]").evaluate(format=Format.FORWARDREF, globals={"glob": dict})
+        self.assertNotIsInstance(generic, ForwardRef)
+        self.assertIs(generic.__origin__, dict)
+        self.assertIs(generic.__args__[0], int)
+
+        generic = ForwardRef("loc[int, undef]").evaluate(format=Format.FORWARDREF, locals={"loc": dict})
+        self.assertNotIsInstance(generic, ForwardRef)
+        self.assertIs(generic.__origin__, dict)
+        self.assertIs(generic.__args__[0], int)
+
+        # Ensure that globals overwrite builtins
+        generic = ForwardRef("list[int, undef]").evaluate(format=Format.FORWARDREF, globals={"list": dict})
+        self.assertNotIsInstance(generic, ForwardRef)
+        self.assertIs(generic.__origin__, dict)
+        self.assertIs(generic.__args__[0], int)
```

Comment:
What about other args?

Reply:
I didn't think it was necessary to check the other args - I'll check them now.